fix: support results with STRUCT and ARRAY columns containing JSON subfields in `to_pandas_batches()`

shuoweil opened the pull request:
This commit addresses issue where creating empty DataFrames with nested JSON columns would fail due to PyArrow's inability to create empty arrays with db_dtypes.JSONArrowType ([Apache Arrow issue #45262](https://github.com/apache/arrow/issues/45262)).  
  
Changes:  
- First tries to create an empty Arrow table directly from the schema
- If that fails with pa.ArrowNotImplementedError, falls back to using storage types (use_storage_types=True)
- Converts the Arrow table to pandas, which properly preserves dtypes

This workaround is specifically needed for the anywidget backend which uses to_pandas_batches() 

Fixes #<456577463> 🦕

shuoweil committed:
Correctly display DataFrames with JSON columns in anywidget

shuoweil committed:
Improve JSON type handling for to_gbq and to_pandas_batches

shuoweil committed:
Revert "Correctly display DataFrames with JSON columns in anywidget"

This reverts commit 8c3451266c28ec0da6dd57c4f9929ae68a593574.

shuoweil committed:
Remove unnecessary comment

tswast reviewed (commented):
Thanks! Look good to me once typing and presubmit failures are addressed.

shuoweil committed:
Merge branch 'main' into shuowei-json-empty-dataframe

shuoweil committed:
code refactor

shuoweil committed:
testcase update

shuoweil committed:
Merge branch 'main' into shuowei-json-empty-dataframe

shuoweil committed:
Fix testcase

shuoweil committed:
Merge branch 'main' into shuowei-json-empty-dataframe

shuoweil committed:
function call updated in
  bigframes/core/blocks.py, unused function removed from bigframes/dtypes.py

shuoweil committed:
revert the code refactor in loader.py, I will use a seperate pr for this refactor

shuoweil commented on bigframes/core/blocks.py:
```diff
@@ -721,6 +721,34 @@ def to_pandas_batches(
+                    # This branch should ideally not be reached if
+                    # contains_db_dtypes_json_dtype is accurate,
+                    # but it's here for MyPy's sake.
+                    series_map[col] = pd.Series([], dtype=dtype)
```
I will reuse this function instead of writing a new one: https://github.com/googleapis/python-bigquery-dataframes/blob/main/bigframes/dtypes.py#L556-L571

shuoweil commented on bigframes/dtypes.py:
```diff
@@ -954,6 +954,21 @@ def contains_db_dtypes_json_dtype(dtype):
     return contains_db_dtypes_json_arrow_type(dtype.pyarrow_dtype)
 
 
+def _replace_json_arrow_with_string(pa_type: pa.DataType) -> pa.DataType:
```
Since I reuse an existing API, this code refactor is no longer relevant to this PR. I will start a new PR for this code refactor.

shuoweil committed:
replace the manual
  construction of the empty DataFrame with the more robust try...except block
  that leverages to_pyarrow and empty_table

shuoweil committed:
fix testcase

shuoweil committed:
existing arrow_to_pandas() helper that properly handles dtype conversion

shuoweil committed:
testcase update

shuoweil committed:
refactor testcase

shuoweil committed:
Add pyarrow id to comments

shuoweil committed:
Merge branch 'main' into shuowei-json-empty-dataframe
